File: app/src/main/java/de/xikolo/controllers/base/BaseActivity.java
package de.xikolo.controllers.base;

import android.content.Intent;
import android.content.pm.PackageManager;
import android.os.Bundle;
import android.os.Handler;
import android.view.KeyEvent;
import android.view.Menu;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;

import androidx.annotation.LayoutRes;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.appcompat.app.ActionBar;
import androidx.appcompat.app.AppCompatActivity;
import androidx.appcompat.widget.Toolbar;
import androidx.coordinatorlayout.widget.CoordinatorLayout;
import androidx.core.content.ContextCompat;
import androidx.drawerlayout.widget.DrawerLayout;

import com.crashlytics.android.Crashlytics;
import com.google.android.gms.cast.framework.CastButtonFactory;
import com.google.android.gms.cast.framework.CastContext;
import com.google.android.gms.cast.framework.CastState;
import com.google.android.gms.cast.framework.CastStateListener;
import com.google.android.gms.cast.framework.IntroductoryOverlay;
import com.google.android.material.appbar.AppBarLayout;
import com.yatatsu.autobundle.AutoBundle;

import butterknife.ButterKnife;
import de.xikolo.App;
import de.xikolo.R;
import de.xikolo.utils.NotificationUtil;
import de.xikolo.utils.extensions.PlayServicesUtil;

public abstract class BaseActivity extends AppCompatActivity implements CastStateListener {

    protected App app;

    protected ActionBar actionBar;

    protected Toolbar toolbar;

    protected AppBarLayout appBar;

    private CastContext castContext;

    private DrawerLayout drawerLayout;

    private CoordinatorLayout contentLayout;

    private MenuItem mediaRouteMenuItem;

    private boolean offlineModeToolbar;

    private IntroductoryOverlay overlay;

    private boolean translucentActionbar = false;

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        if (savedInstanceState != null) {
            // restore
            AutoBundle.bind(this, savedInstanceState);
        } else {
            AutoBundle.bind(this);
        }

        app = App.getInstance();

        offlineModeToolbar = true;

        try {
            if (PlayServicesUtil.getHasPlayServices(this)) {
                castContext = CastContext.getSharedInstance(this);
            }
        } catch (Exception e) {
            Crashlytics.logException(e);
        }

        if (overlay == null) {
            showOverlay();
        }

        handleIntent(getIntent());

        App.getInstance().getState().getConnectivity().observe(this, this::onConnectivityChange);
    }

    @Override
    public void setContentView(@LayoutRes int layoutResID) {
        super.setContentView(layoutResID);
        ButterKnife.bind(this, findViewById(android.R.id.content));
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        super.onCreateOptionsMenu(menu);

        try {
            if (PlayServicesUtil.getHasPlayServices(this)) {
                castContext = CastContext.getSharedInstance(this);
                getMenuInflater().inflate(R.menu.cast, menu);
                mediaRouteMenuItem = CastButtonFactory.setUpMediaRouteButton(
                        getApplicationContext(),
                        menu,
                        R.id.media_route_menu_item);
            }
        } catch (Exception e) {
            Crashlytics.logException(e);
        }

        return true;
    }

    @Override
    protected void onNewIntent(Intent intent) {
        super.onNewIntent(intent);

        setIntent(intent);
        AutoBundle.bind(this);

        handleIntent(intent);
    }

    @Override
    protected void onStart() {
        super.onStart();

        if (castContext != null) {
            castContext.addCastStateListener(this);
            setupCastMiniController();
        }
    }

    @Override
    protected void onPause() {
        super.onPause();
        app.syncCookieSyncManager();
    }

    @Override
    protected void onStop() {
        super.onStop();

        if (castContext != null) {
            castContext.removeCastStateListener(this);
        }
    }

    @Override
    protected void onSaveInstanceState(Bundle outState) {
        super.onSaveInstanceState(outState);
        AutoBundle.pack(this, outState);
    }

    private void showOverlay() {
        if (overlay != null) {
            overlay.remove();
        }
        if ((mediaRouteMenuItem != null) && mediaRouteMenuItem.isVisible()) {
            new Handler().postDelayed(() -> {
                if (mediaRouteMenuItem != null && mediaRouteMenuItem.isVisible()) {
                    overlay = new IntroductoryOverlay.Builder(BaseActivity.this, mediaRouteMenuItem)
                            .setTitleText(R.string.intro_overlay_text)
                            .setSingleTime()
                            .setOnOverlayDismissedListener(() -> overlay = null)
                            .build();
                    overlay.show();
                }
            }, 1000);
        }
    }

    @Override
    public void onCastStateChanged(int newState) {
        if (newState != CastState.NO_DEVICES_AVAILABLE) {
            showOverlay();
        }
    }

    protected void setupActionBar() {
        setupActionBar(false);
    }

    protected void setupActionBar(boolean translucentActionbar) {
        this.translucentActionbar = translucentActionbar;
        Toolbar tb = findViewById(R.id.toolbar);
        if (tb != null) {
            setSupportActionBar(tb);
        }

        toolbar = tb;

        actionBar = getSupportActionBar();

        if (actionBar != null) {
            actionBar.setDisplayHomeAsUpEnabled(true);
            actionBar.setHomeButtonEnabled(true);
        }

        drawerLayout = findViewById(R.id.drawer_layout);
        contentLayout = findViewById(R.id.contentLayout);
        appBar = findViewById(R.id.appbar);
        setColorScheme(R.color.toolbar, R.color.statusbar);
    }

    private boolean setupCastMiniController() {
        if (PlayServicesUtil.getHasPlayServices(this) && findViewById(R.id.miniControllerContainer) != null) {
            View container = findViewById(R.id.miniControllerContainer);
            ViewGroup parent = (ViewGroup) container.getParent();
            int index = parent.indexOfChild(container);

            parent.removeView(container);
            container = getLayoutInflater().inflate(R.layout.container_mini_controller, parent, false);
            parent.addView(container, index);

            return true;
        } else {
            return false;
        }
    }

    @SuppressWarnings("unused")
    protected void setActionBarElevation(float elevation) {
        if (actionBar != null) {
            actionBar.setElevation(elevation);
        }
    }

    protected void setAppBarExpanded(boolean expanded) {
        if (appBar != null) {
            appBar.setExpanded(expanded, false);
        }
    }

    protected void enableOfflineModeToolbar(boolean enable) {
        this.offlineModeToolbar = enable;
    }

    protected void setColorScheme(int toolbarColor, int statusbarColor) {
        if (toolbar != null && !translucentActionbar) {
            toolbar.setBackgroundColor(ContextCompat.getColor(this, toolbarColor));
            if (drawerLayout != null) {
                drawerLayout.setStatusBarBackgroundColor(ContextCompat.getColor(this, toolbarColor));
            } else {
                getWindow().setStatusBarColor(ContextCompat.getColor(this, statusbarColor));
            }
            if (contentLayout != null) {
                contentLayout.setBackgroundColor(ContextCompat.getColor(this, toolbarColor));
            }
        }
    }

    public void onConnectivityChange(boolean isOnline) {
        if (toolbar != null && offlineModeToolbar) {
            if (isOnline) {
                toolbar.setSubtitle("");
                setColorScheme(R.color.toolbar, R.color.statusbar);
            } else {
                toolbar.setSubtitle(getString(R.string.offline_mode));
                setColorScheme(R.color.toolbar_offline, R.color.statusbar_offline);
            }
        }
    }

    @Override
    public void onRequestPermissionsResult(int requestCode, @NonNull String permissions[], @NonNull int[] grantResults) {
        if (grantResults.length > 0
                && grantResults[0] == PackageManager.PERMISSION_GRANTED) {
            App.getInstance().getState().getPermission().of(requestCode).granted();
        } else {
            App.getInstance().getState().getPermission().of(requestCode).denied();
        }
    }

    protected void enableCastMediaRouterButton(boolean enable) {
        if (mediaRouteMenuItem != null) {
            mediaRouteMenuItem.setVisible(enable);
            invalidateOptionsMenu();
        }
    }

    private void handleIntent(Intent intent) {
        NotificationUtil.deleteDownloadNotificationsFromIntent(intent);
    }

    @Override
    public boolean dispatchKeyEvent(KeyEvent event) {
        if (PlayServicesUtil.getHasPlayServices(this) && castContext != null) {
            return castContext.onDispatchVolumeKeyEventBeforeJellyBean(event)
                    || super.dispatchKeyEvent(event);
        } else {
            return super.dispatchKeyEvent(event);
        }
    }

}

File: app/src/main/java/de/xikolo/controllers/base/BaseActivity.kt
package de.xikolo.controllers.base

import android.content.Intent
import android.content.pm.PackageManager
import android.os.Bundle
import android.os.Handler
import android.view.KeyEvent
import android.view.Menu
import android.view.MenuItem
import android.view.View
import android.view.ViewGroup
import androidx.annotation.LayoutRes
import androidx.appcompat.app.ActionBar
import androidx.appcompat.app.AppCompatActivity
import androidx.appcompat.widget.Toolbar
import androidx.coordinatorlayout.widget.CoordinatorLayout
import androidx.core.content.ContextCompat
import androidx.drawerlayout.widget.DrawerLayout
import butterknife.ButterKnife
import com.google.android.gms.cast.framework.CastButtonFactory
import com.google.android.gms.cast.framework.CastContext
import com.google.android.gms.cast.framework.CastState
import com.google.android.gms.cast.framework.CastStateListener
import com.google.android.gms.cast.framework.IntroductoryOverlay
import com.google.android.material.appbar.AppBarLayout
import com.google.android.material.appbar.AppBarLayout.LayoutParams.SCROLL_FLAG_ENTER_ALWAYS
import com.google.android.material.appbar.AppBarLayout.LayoutParams.SCROLL_FLAG_SCROLL
import com.google.android.material.appbar.AppBarLayout.LayoutParams.SCROLL_FLAG_SNAP
import com.google.firebase.crashlytics.FirebaseCrashlytics
import com.yatatsu.autobundle.AutoBundle
import de.xikolo.App
import de.xikolo.R
import de.xikolo.extensions.observe
import de.xikolo.utils.NotificationUtil
import de.xikolo.utils.extensions.hasPlayServices

abstract class BaseActivity : AppCompatActivity(), CastStateListener {

    protected var actionBar: ActionBar? = null

    protected var toolbar: Toolbar? = null

    protected var appBar: AppBarLayout? = null

    private var castContext: CastContext? = null

    private var drawerLayout: DrawerLayout? = null

    private var contentLayout: CoordinatorLayout? = null

    private var mediaRouteMenuItem: MenuItem? = null

    private var offlineModeToolbar: Boolean = false

    private var overlay: IntroductoryOverlay? = null

    private var translucentActionbar = false

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        if (savedInstanceState != null) {
            // restore
            AutoBundle.bind(this, savedInstanceState)
        } else {
            AutoBundle.bind(this)
        }

        offlineModeToolbar = true

        try {
            if (this.hasPlayServices) {
                castContext = CastContext.getSharedInstance(this)
            }
        } catch (e: Exception) {
            FirebaseCrashlytics.getInstance().recordException(e)
        }

        if (overlay == null) {
            showOverlay()
        }

        handleIntent(intent)

        App.instance.state.connectivity.observe(this) {
            this.onConnectivityChange(it)
        }
    }

    override fun setContentView(@LayoutRes layoutResID: Int) {
        super.setContentView(layoutResID)
        ButterKnife.bind(this, findViewById<View>(android.R.id.content))
    }

    override fun onCreateOptionsMenu(menu: Menu?): Boolean {
        super.onCreateOptionsMenu(menu)

        try {
            if (this.hasPlayServices) {
                castContext = CastContext.getSharedInstance(this)
                menuInflater.inflate(R.menu.cast, menu)
                mediaRouteMenuItem = CastButtonFactory.setUpMediaRouteButton(
                    applicationContext,
                    menu,
                    R.id.media_route_menu_item)
            }
        } catch (e: Exception) {
            FirebaseCrashlytics.getInstance().recordException(e)
        }

        return true
    }

    override fun onNewIntent(intent: Intent?) {
        super.onNewIntent(intent)

        setIntent(intent)
        AutoBundle.bind(this)

        handleIntent(intent)
    }

    override fun onStart() {
        super.onStart()

        if (castContext != null) {
            castContext?.addCastStateListener(this)
            setupCastMiniController()
        }
    }

    override fun onPause() {
        super.onPause()
        App.instance.syncCookieSyncManager()
    }

    override fun onStop() {
        super.onStop()

        castContext?.removeCastStateListener(this)
    }

    override fun onSaveInstanceState(outState: Bundle) {
        super.onSaveInstanceState(outState)
        AutoBundle.pack(this, outState)
    }

    private fun showOverlay() {
        overlay?.remove()
        if (mediaRouteMenuItem?.isVisible == true) {
            Handler().postDelayed({
                if (mediaRouteMenuItem != null && mediaRouteMenuItem?.isVisible == true) {
                    overlay = IntroductoryOverlay.Builder(this@BaseActivity, mediaRouteMenuItem)
                        .setTitleText(R.string.intro_overlay_text)
                        .setSingleTime()
                        .setOnOverlayDismissedListener { overlay = null }
                        .build()
                    overlay?.show()
                }
            }, 1000)
        }
    }

    override fun onCastStateChanged(newState: Int) {
        if (newState != CastState.NO_DEVICES_AVAILABLE) {
            showOverlay()
        }
    }

    @JvmOverloads
    protected fun setupActionBar(translucentActionbar: Boolean = false) {
        this.translucentActionbar = translucentActionbar
        val tb = findViewById<Toolbar>(R.id.toolbar)
        if (tb != null) {
            setSupportActionBar(tb)
        }

        toolbar = tb

        actionBar = supportActionBar
        actionBar?.setDisplayHomeAsUpEnabled(true)
        actionBar?.setHomeButtonEnabled(true)

        drawerLayout = findViewById(R.id.drawer_layout)
        contentLayout = findViewById(R.id.contentLayout)
        appBar = findViewById(R.id.appbar)
        setColorScheme(R.color.toolbar, R.color.statusbar)
    }

    private fun setupCastMiniController(): Boolean {
        return if (this.hasPlayServices && findViewById<View>(R.id.miniControllerContainer) != null) {
            var container = findViewById<View>(R.id.miniControllerContainer)
            val parent = container.parent as ViewGroup
            val index = parent.indexOfChild(container)

            parent.removeView(container)
            container = layoutInflater.inflate(R.layout.container_mini_controller, parent, false)
            parent.addView(container, index)

            true
        } else {
            false
        }
    }

    protected fun setActionBarElevation(elevation: Float) {
        actionBar?.elevation = elevation
    }

    protected fun setAppBarExpanded(expanded: Boolean) {
        appBar?.setExpanded(expanded, false)
    }

    fun setScrollingBehavior(hide: Boolean) {
        toolbar?.layoutParams = (toolbar?.layoutParams as? AppBarLayout.LayoutParams)?.apply {
            scrollFlags = SCROLL_FLAG_ENTER_ALWAYS or SCROLL_FLAG_SNAP or if (hide) SCROLL_FLAG_SCROLL else 0
        }
    }

    protected fun enableOfflineModeToolbar(enable: Boolean) {
        this.offlineModeToolbar = enable
    }

    protected fun setColorScheme(toolbarColor: Int, statusbarColor: Int) {
        if (!translucentActionbar) {
            toolbar?.let { toolbar ->
                toolbar.setBackgroundColor(ContextCompat.getColor(this, toolbarColor))

                drawerLayout?.setStatusBarBackgroundColor(ContextCompat.getColor(this, toolbarColor))
                if (drawerLayout == null) {
                    window.statusBarColor = ContextCompat.getColor(this, statusbarColor)
                }

                contentLayout?.setBackgroundColor(ContextCompat.getColor(this, toolbarColor))
            }
        }
    }

    open fun onConnectivityChange(isOnline: Boolean) {
        if (offlineModeToolbar) {
            toolbar?.let {
                if (isOnline) {
                    it.subtitle = ""
                    setColorScheme(R.color.toolbar, R.color.statusbar)
                } else {
                    it.subtitle = getString(R.string.offline_mode)
                    setColorScheme(R.color.toolbar_offline, R.color.statusbar_offline)
                }
            }
        }
    }

    override fun onRequestPermissionsResult(requestCode: Int, permissions: Array<out String>, grantResults: IntArray) {
        if (grantResults.isNotEmpty() && grantResults[0] == PackageManager.PERMISSION_GRANTED) {
            App.instance.state.permission.of(requestCode).granted()
        } else {
            App.instance.state.permission.of(requestCode).denied()
        }
    }

    protected fun enableCastMediaRouterButton(enable: Boolean) {
        mediaRouteMenuItem?.isVisible = enable
        invalidateOptionsMenu()
    }

    private fun handleIntent(intent: Intent?) {
        if (intent != null) {
            NotificationUtil.getInstance(this).deleteDownloadNotificationsFromIntent(intent)
        }
    }

    override fun dispatchKeyEvent(event: KeyEvent): Boolean {
        return if (this.hasPlayServices && castContext != null) {
            castContext?.onDispatchVolumeKeyEventBeforeJellyBean(event) == true || super.dispatchKeyEvent(event)
        } else {
            super.dispatchKeyEvent(event)
        }
    }

}
